contrib/colab/DC-GAN.ipynb

```diff
@@ -0,0 +1,554 @@
+        "        \n",
+        "\n",
+        "        # 2. Train on Fake Data\n",
+        "        prediction_fake = discriminator(fake_data)\n",
+        "        # Calculate error and backpropagate\n",
+        "\n",
+        "        error_fake = loss(prediction_fake, fake_data_target(real_data.size(0), device))\n",
+        "        total_error = error_real + error_fake\n",
+        "        total_error.backward()\n",
+        "\n",
+        "        # Update weights with gradients\n",
+        "        xm.optimizer_step(optimizer)\n",
+        "\n",
+        "        return total_error, prediction_real, prediction_fake\n",
+        "\n",
+        "    def train_step_generator(optimizer, fake_data, device):\n",
+        "        # Reset gradients\n",
+        "        optimizer.zero_grad()\n",
+        "        prediction = discriminator(fake_data)\n",
+        "        # Calculate error and backpropagate\n",
+        "        error = loss(prediction, real_data_target(prediction.size(0), device))\n",
+        "        error.backward()\n",
+        "        # Update weights with gradients\n",
+        "        xm.optimizer_step(optimizer)\n",
+        "\n",
+        "        # Return error\n",
+        "        return error\n",
+        "\n",
+        "    # Notice the use of .detach() when fake_data is to passed into discriminator\n",
+        "    def train_loop_fn(loader):\n",
+        "        tracker = xm.RateTracker()\n",
+        "        for n_batch, (real_batch,_) in enumerate(loader):\n",
+        "            # Train Step Descriminator\n",
+        "            real_data = real_batch.to(device)\n",
+        "            # sample noise and generate fake data\n",
+        "            noise = generator.generate_noise(real_data.size(0)).to(device)\n",
+        "            fake_data = generator(noise)\n",
+        "            d_error, d_pred_real, d_pred_fake = train_step_discriminator(\n",
+        "                d_optimizer, real_data, fake_data.detach(), device)\n",
+        "            \n",
+        "            #Train Step Generator\n",
+        "            noise = generator.generate_noise(real_data.size(0)).to(device)\n",
+        "            fake_data = generator(noise)\n",
+        "            g_error = train_step_generator(g_optimizer, fake_data, device)\n",
+        "        return d_error.item(), g_error.item()\n",
+        "\n",
+        "\n",
+        "    for epoch in range(1, FLAGS['num_epochs'] +1):\n",
+        "        d_error, g_error = train_loop_fn (pl.MpDeviceLoader(train_loader, device))\n",
+        "        xm.master_print(\"Finished training epoch {}: D_error:{}, G_error: {}\".format(epoch, d_error, g_error))\n",
+        "        \n",
+        "    num_test_samples = 24\n",
+        "    test_noise = generator.generate_noise(num_test_samples).to(device)\n",
+        "    xm.do_on_ordinals(plot_results, generator(test_noise).detach().cpu(), (0,))"
```

Comment:
No `cpu()` call in there.
The `xm.do_on_ordinals()` API takes care of that, and the non-executing cores do not need to fetch device data.
Which is particularly critical on memory starved environments like Colab.

Reply:
Thanks Davide. Removed it.